lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+	// more than once in the mempool transactions.
+	txn := server.blockchain.db.NewTransaction(true)
+	defer txn.Discard()
+
+	err = mempoolUtxoView.FlushToDbWithTxn(txn, uint64(server.blockchain.bestChain[len(server.blockchain.bestChain)-1].Height))
+
+	if err != nil || originalCommittedFlushId != stateChangeSyncer.BlockSyncFlushId {
+		mempoolUtxoView.EventManager.stateSyncerFlushed(&StateSyncerFlushedEvent{
+			FlushId:        uuid.Nil,
+			Succeeded:      false,
+			IsMempoolFlush: true,
+		})
+		return false, errors.Wrapf(err, "StateChangeSyncer.SyncMempoolToStateSyncer: ")
+	}
+	mempoolUtxoView.EventManager.stateSyncerFlushed(&StateSyncerFlushedEvent{
+		FlushId:          uuid.Nil,
+		Succeeded:        true,
+		IsMempoolFlush:   true,
+		BlockSyncFlushId: originalCommittedFlushId,
+	})
+
+	// Loop through all the unconnected transactions in the mempool and connect them and their utxo ops to the mempool view.
+	for txHash, unconnectedTxn := range server.mempool.unconnectedTxns {
+		utxoOpsForTxn, _, _, _, err := mempoolUtxoView.ConnectTransaction(
+			unconnectedTxn.tx, &txHash, 0, uint32(blockHeight), false, false /*ignoreUtxos*/)
+		if err != nil {
+			return false, errors.Wrapf(err, "StateChangeSyncer.SyncMempoolToStateSyncer ConnectTransaction: ")
+		}
+
+		// Emit transaction state change.
+		mempoolUtxoView.EventManager.stateSyncerOperation(&StateSyncerOperationEvent{
+			StateChangeEntry: &StateChangeEntry{
+				OperationType: DbOperationTypeUpsert,
+				KeyBytes:      TxnHashToTxnKey(&txHash),
+				EncoderBytes:  EncodeToBytes(blockHeight, unconnectedTxn.tx, false),
+			},
+			FlushId:      uuid.Nil,
+			IsMempoolTxn: true,
+		})
+
+		// Capture the utxo ops for the transaction in a UTXOOp bundle.
+		utxoOpBundle := &UtxoOperationBundle{
+			UtxoOpBundle: [][]*UtxoOperation{},
+		}
+
+		utxoOpBundle.UtxoOpBundle = append(utxoOpBundle.UtxoOpBundle, utxoOpsForTxn)
+
+		// Emit UTXOOp bundle event
+		mempoolUtxoView.EventManager.stateSyncerOperation(&StateSyncerOperationEvent{
+			StateChangeEntry: &StateChangeEntry{
+				OperationType: DbOperationTypeUpsert,
+				KeyBytes:      _DbKeyForTxnUtxoOps(&txHash),
+				EncoderBytes:  EncodeToBytes(blockHeight, utxoOpBundle, false),
+			},
+			FlushId:      uuid.Nil,
+			IsMempoolTxn: true,
+		})
+	}
+
+	return false, nil
+}
+
+func (stateChangeSyncer *StateChangeSyncer) StartMempoolSyncRoutine(server *Server) {
+	go func() {
+		// Wait for mempool to be initialized.
+		for server.mempool == nil || server.blockchain.chainState() != SyncStateFullyCurrent {
+			time.Sleep(1000 * time.Millisecond)
+		}
+		if !stateChangeSyncer.BlocksyncCompleteEntriesFlushed && stateChangeSyncer.SyncType == NodeSyncTypeBlockSync {
+			stateChangeSyncer.FlushAllEntriesToFile(server)
+		}
+		mempoolClosed := server.mempool.stopped
+		for !mempoolClosed {
+			// Sleep for a short while to avoid a tight loop.
+			time.Sleep(1000 * time.Millisecond)
```

Comment:
Wonder how low we can make this value. Would be cool if you could submit a txn and have it show up in PG in like 100ms. I'd maybe bias toward a lower value until we run into performance issues.